File: algorithm/543.diameter-of-binary-tree/543.diameter-of-binary-tree.java
/*
 * @lc app=leetcode id=543 lang=java
 *
 * [543] Diameter of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 /**
  * I didn't get it right the first time though my reasoning is correct.
  * loop through the all the node and find the one which has biggest sum of depth of left and right
  * 
  * The provided method itself don't have to be the main recursive method
  */
class Solution {
    int max = Integer.MIN_VALUE;
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null) return 0;
        depthOfBT(root);
        return max;
    }
    private int depthOfBT(TreeNode root) {
        if(root == null) return 0;
        int left = depthOfBT(root.left);
        int right = depthOfBT(root.right);
        max = Math.max(max, left + right);
        return Math.max(left, right) + 1;
    }


}
// @lc code=end

File: algorithm/543.diameter-of-binary-tree/543.diameter-of-binary-tree.py
#
# @lc app=leetcode id=543 lang=python3
#
# [543] Diameter of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.ans = 0

        def get_height(root: Optional[TreeNode]) -> int:
            if root is None:
                return 0
            left_h = get_height(root.left)
            right_h = get_height(root.right)
            # 穿过当前节点的直径 = 左子树高度 + 右子树高度
            self.ans = max(self.ans, left_h + right_h)
            return max(left_h, right_h) + 1

        get_height(root)
        return self.ans
# @lc code=end
